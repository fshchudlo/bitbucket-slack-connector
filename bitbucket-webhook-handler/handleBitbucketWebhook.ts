import * as useCases from "./use-cases";
import { SlackAPIAdapter } from "./ports/SlackAPIAdapter";
import { BitbucketAPIAdapter } from "./ports/BitbucketAPIAdapter";
import { BitbucketNotification } from "../bitbucket-payload-types";
import { WebhookHandlerConfig } from "./webhookHandlerConfig";
import { provisionPullRequestChannel } from "./provisionPullRequestChannel";

export default async function handleBitbucketWebhook(payload: BitbucketNotification, slackAPI: SlackAPIAdapter, bitbucketAPI: BitbucketAPIAdapter, config: WebhookHandlerConfig) {
    const eventKey = payload.eventKey;
    const channelInfo = await provisionPullRequestChannel(slackAPI, bitbucketAPI, payload, config);

    switch (eventKey) {
        case "pr:opened":
            await useCases.inviteParticipantsAndSetChannelBookmark(payload, slackAPI, config.defaultChannelParticipants, channelInfo.id);
            await useCases.tryBroadcastMessageAboutOpenedPR(payload, slackAPI, config.getOpenedPRBroadcastChannelId(payload));
            break;
        case "pr:modified":
            await useCases.sendMessageAboutPRModification(payload, slackAPI, channelInfo.id);
            break;
        case "pr:reviewer:updated":
            await useCases.updateChannelMembers(payload, slackAPI, channelInfo.id);
            break;
        case "pr:reviewer:unapproved":
        case "pr:reviewer:needs_work":
        case "pr:reviewer:approved":
            await useCases.sendMessageAboutReviewerAction(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:added":
            await useCases.sendMessageAboutAddedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:edited":
            await useCases.sendMessageAboutEditedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:deleted":
            await useCases.sendMessageAboutDeletedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:from_ref_updated":
            await useCases.sendMessageAboutNewCommit(payload, slackAPI, bitbucketAPI, channelInfo.id);
            break;
        case "pr:merged":
        case "pr:declined":
        case "pr:deleted":
            await useCases.sendCompletionMessageAndCloseTheChannel(payload, slackAPI, channelInfo.id);
            await useCases.tryBroadcastMessageAboutClosedPR(payload, slackAPI, config.getOpenedPRBroadcastChannelId(payload));
            break;
        default:
            throw new Error(`"${eventKey}" event key is unknown.`);
    }
}

