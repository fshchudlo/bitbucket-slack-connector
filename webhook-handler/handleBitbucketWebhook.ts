import {
    sendCompletionMessageAndCloseTheChannel,
    setChannelTopicAndInviteParticipants,
    sendMessageAboutAddedComment,
    updateChannelMembers,
    sendMessageAboutNewCommit,
    sendMessageAboutReviewerAction,
    sendMessageAboutDeletedComment, sendMessageAboutEditedComment
} from "./use-cases";
import { SlackAPIAdapter } from "./SlackAPIAdapter";
import { BitbucketGateway } from "./BitbucketGateway";
import { sendMessageAboutPRModification } from "./use-cases/sendMessageAboutPRModification";
import { BitbucketNotification } from "../typings";
import { buildChannelName } from "./slack-helpers";

export default async function handleBitbucketWebhook(payload: BitbucketNotification, slackAPI: SlackAPIAdapter, bitbucketGateway: BitbucketGateway, usePrivateChannels: boolean = true) {
    const eventKey = payload.eventKey;
    const channelInfo = await provisionPullRequestChannel(slackAPI, payload, usePrivateChannels);
    switch (eventKey) {
        case "pr:opened":
            await setChannelTopicAndInviteParticipants(payload, slackAPI, channelInfo);
            break;
        case "pr:modified":
            await sendMessageAboutPRModification(payload, slackAPI, channelInfo);
            break;
        case "pr:reviewer:updated":
            await updateChannelMembers(payload, slackAPI, channelInfo);
            break;
        case "pr:reviewer:unapproved":
        case "pr:reviewer:needs_work":
        case "pr:reviewer:approved":
            await sendMessageAboutReviewerAction(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:added":
            await sendMessageAboutAddedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:edited":
            await sendMessageAboutEditedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:deleted":
            await sendMessageAboutDeletedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:from_ref_updated":
            await sendMessageAboutNewCommit(payload, slackAPI, bitbucketGateway, channelInfo);
            break;
        case "pr:merged":
        case "pr:declined":
        case "pr:deleted":
            await sendCompletionMessageAndCloseTheChannel(payload, slackAPI, channelInfo);
            break;
        default:
            throw new Error(`"${eventKey}" event key is unknown.`);
    }
}

async function provisionPullRequestChannel(slackAPI: SlackAPIAdapter, payload: BitbucketNotification, usePrivateChannels: boolean) {
    return payload.eventKey == "pr:opened" ? await slackAPI.createChannel({
        name: buildChannelName(payload.pullRequest),
        isPrivate: usePrivateChannels
    }) : await slackAPI.findChannel(buildChannelName(payload.pullRequest), true);
}
