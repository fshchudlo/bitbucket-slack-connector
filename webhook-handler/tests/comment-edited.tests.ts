import SlackAdapterSnapshottingMock from "./mocks/SlackAdapterSnapshottingMock";
import TestPayloadBuilder from "./mocks/TestPayloadBuilder";
import handleBitbucketWebhook from "../handleBitbucketWebhook";
import { TestBitbucketGateway } from "./mocks/TestBitbucketGateway";

describe("handleBitbucketWebhook", () => {
    it("Should send message on PR comment edit", async () => {
        const testSlackGateway = new SlackAdapterSnapshottingMock();
        const payload = TestPayloadBuilder.pullRequestCommentEdited();
        await handleBitbucketWebhook(payload, testSlackGateway, new TestBitbucketGateway());

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should send message on comment conversion to the task", async () => {
        const testSlackGateway = new SlackAdapterSnapshottingMock();
        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestCommentAdded(), testSlackGateway, new TestBitbucketGateway());


        const payload = TestPayloadBuilder.pullRequestCommentConvertedToTheTask();
        await handleBitbucketWebhook(payload, testSlackGateway, new TestBitbucketGateway());

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should send message on task conversion to the comment", async () => {
        const testSlackGateway = new SlackAdapterSnapshottingMock();
        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestTaskAdded(), testSlackGateway, new TestBitbucketGateway());


        const payload = TestPayloadBuilder.pullRequestTaskConvertedToTheComment();
        await handleBitbucketWebhook(payload, testSlackGateway, new TestBitbucketGateway());

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });
    it("Should send message on comment resolving and reopening", async () => {
        const testSlackGateway = new SlackAdapterSnapshottingMock();
        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestCommentAdded(), testSlackGateway, new TestBitbucketGateway());


        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestCommentResolved(), testSlackGateway, new TestBitbucketGateway());

        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestCommentReopened(), testSlackGateway, new TestBitbucketGateway());

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });
    it("Should send message on task resolving and reopening", async () => {
        const testSlackGateway = new SlackAdapterSnapshottingMock();
        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestTaskAdded(), testSlackGateway, new TestBitbucketGateway());


        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestTaskResolved(), testSlackGateway, new TestBitbucketGateway());

        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestTaskReopened(), testSlackGateway, new TestBitbucketGateway());

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });
});
