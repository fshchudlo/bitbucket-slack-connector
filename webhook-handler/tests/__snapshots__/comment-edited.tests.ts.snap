// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleBitbucketWebhook Should return generic message if initial comment was not found 1`] = `
{
  "addedBookmarks": [],
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "searchedChannels": [
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
  ],
  "searchedCommentSnapshots": [
    {
      "bitbucketCommentId": 1,
      "channelId": "12345",
    },
  ],
  "sentMessages": [
    {
      "blocks": [
        {
          "text": {
            "text": ":writing_hand: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|edited task>:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Test comment",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "BLOCKER",
          "taskResolvedDate": undefined,
          "threadResolvedDate": undefined,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": undefined,
      "text": ":writing_hand: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|edited task>:",
      "threadId": undefined,
    },
  ],
}
`;

exports[`handleBitbucketWebhook Should send message on PR comment edit 1`] = `
{
  "addedBookmarks": [],
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "searchedChannels": [
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
  ],
  "searchedCommentSnapshots": [
    {
      "bitbucketCommentId": 1,
      "channelId": "12345",
    },
  ],
  "sentMessages": [
    {
      "blocks": [
        {
          "text": {
            "text": ":loudspeaker: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|added comment>:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Test comment",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "NORMAL",
          "taskResolvedDate": undefined,
          "threadResolvedDate": undefined,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": undefined,
      "text": ":loudspeaker: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|added comment>:",
      "threadId": undefined,
    },
    {
      "blocks": [
        {
          "text": {
            "text": ":writing_hand: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|edited comment>:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Updated comment text",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "NORMAL",
          "taskResolvedDate": undefined,
          "threadResolvedDate": undefined,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": true,
      "text": ":writing_hand: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|edited comment>:",
      "threadId": "ABCDE",
    },
  ],
}
`;

exports[`handleBitbucketWebhook Should send message on comment conversion to the task 1`] = `
{
  "addedBookmarks": [],
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "searchedChannels": [
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
  ],
  "searchedCommentSnapshots": [
    {
      "bitbucketCommentId": 1,
      "channelId": "12345",
    },
  ],
  "sentMessages": [
    {
      "blocks": [
        {
          "text": {
            "text": ":loudspeaker: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|added comment>:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Test comment",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "NORMAL",
          "taskResolvedDate": undefined,
          "threadResolvedDate": undefined,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": undefined,
      "text": ":loudspeaker: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|added comment>:",
      "threadId": undefined,
    },
    {
      "blocks": [
        {
          "text": {
            "text": ":pushpin: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|converted comment to the task>:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Test comment",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "BLOCKER",
          "taskResolvedDate": undefined,
          "threadResolvedDate": undefined,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": true,
      "text": ":pushpin: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|converted comment to the task>:",
      "threadId": "ABCDE",
    },
  ],
}
`;

exports[`handleBitbucketWebhook Should send message on comment resolving and reopening 1`] = `
{
  "addedBookmarks": [],
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "searchedChannels": [
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
  ],
  "searchedCommentSnapshots": [
    {
      "bitbucketCommentId": 1,
      "channelId": "12345",
    },
    {
      "bitbucketCommentId": 1,
      "channelId": "12345",
    },
  ],
  "sentMessages": [
    {
      "blocks": [
        {
          "text": {
            "text": ":loudspeaker: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|added comment>:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Test comment",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "NORMAL",
          "taskResolvedDate": undefined,
          "threadResolvedDate": undefined,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": undefined,
      "text": ":loudspeaker: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|added comment>:",
      "threadId": undefined,
    },
    {
      "blocks": [
        {
          "text": {
            "text": ":white_check_mark: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|resolved comment>:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Test comment",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "NORMAL",
          "taskResolvedDate": undefined,
          "threadResolvedDate": 123456789,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": true,
      "text": ":white_check_mark: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|resolved comment>:",
      "threadId": "ABCDE",
    },
    {
      "blocks": [
        {
          "text": {
            "text": ":repeat: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|reopened comment>:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Test comment",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "NORMAL",
          "taskResolvedDate": undefined,
          "threadResolvedDate": undefined,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": true,
      "text": ":repeat: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|reopened comment>:",
      "threadId": "ABCDE",
    },
  ],
}
`;

exports[`handleBitbucketWebhook Should send message on task conversion to the comment 1`] = `
{
  "addedBookmarks": [],
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "searchedChannels": [
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
  ],
  "searchedCommentSnapshots": [
    {
      "bitbucketCommentId": 1,
      "channelId": "12345",
    },
  ],
  "sentMessages": [
    {
      "blocks": [
        {
          "text": {
            "text": ":loudspeaker: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|added task>:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Test comment",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "BLOCKER",
          "taskResolvedDate": undefined,
          "threadResolvedDate": undefined,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": undefined,
      "text": ":loudspeaker: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|added task>:",
      "threadId": undefined,
    },
    {
      "blocks": [
        {
          "text": {
            "text": ":writing_hand: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|converted task to the comment>:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Test comment",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "NORMAL",
          "taskResolvedDate": undefined,
          "threadResolvedDate": undefined,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": true,
      "text": ":writing_hand: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|converted task to the comment>:",
      "threadId": "ABCDE",
    },
  ],
}
`;

exports[`handleBitbucketWebhook Should send message on task resolving and reopening 1`] = `
{
  "addedBookmarks": [],
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "searchedChannels": [
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
  ],
  "searchedCommentSnapshots": [
    {
      "bitbucketCommentId": 1,
      "channelId": "12345",
    },
    {
      "bitbucketCommentId": 1,
      "channelId": "12345",
    },
  ],
  "sentMessages": [
    {
      "blocks": [
        {
          "text": {
            "text": ":loudspeaker: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|added task>:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Test comment",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "BLOCKER",
          "taskResolvedDate": undefined,
          "threadResolvedDate": undefined,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": undefined,
      "text": ":loudspeaker: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|added task>:",
      "threadId": undefined,
    },
    {
      "blocks": [
        {
          "text": {
            "text": ":white_check_mark: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|resolved task>:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Test comment",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "BLOCKER",
          "taskResolvedDate": 123456789,
          "threadResolvedDate": undefined,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": true,
      "text": ":white_check_mark: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|resolved task>:",
      "threadId": "ABCDE",
    },
    {
      "blocks": [
        {
          "text": {
            "text": ":repeat: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|reopened task>:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Test comment",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "BLOCKER",
          "taskResolvedDate": undefined,
          "threadResolvedDate": undefined,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": true,
      "text": ":repeat: Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|reopened task>:",
      "threadId": "ABCDE",
    },
  ],
}
`;
